File: examples/deployed/go/check.go
// Ask the deployed head node the same questions the Python check asks, through
// the Go SDK, and require the same answers.
//
// The oracle stays in Python. `check.py --expect` writes what its fold
// computed and this reads it, rather than decoding the packed trip file again:
// three decoders of one file would be three places to be wrong, and a
// common-mode error in them would agree with itself. What is being checked
// here is the *client* — that a Go caller building the same request over the
// same socket gets the same answer as a Python one.
//
// Which matters because the two build requests differently. The Python client
// declares a `Table` and carries a schema fingerprint; this one names a table
// and an ordinal. A join's computed value is `join.computed(0)` there and
// `slate.JoinComputed(0)` here. Every one of those is a place a client could
// send something subtly different, and until this file the deployed example
// exercised exactly one of them.
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"math"
	"os"
	"sort"
	"time"

	"github.com/howlerops/slate-orm/clients/go/slate"
)

// The ordinals `head.toml` declares. Written out rather than fetched, which is
// the Go client's arrangement: it names a table and an ordinal, and the server
// has no way to tell it that `3` is `pickup_time` — which is why the Python
// client's fingerprint check exists and why these constants are worth naming.
const (
	tripID         slate.Ordinal = 0
	tripPickupZone slate.Ordinal = 1
	tripPickupTime slate.Ordinal = 3
	tripFare       slate.Ordinal = 7
	zoneID         slate.Ordinal = 0
	zoneBorough    slate.Ordinal = 1
)

// expected is what `check.py --expect` wrote.
type expected struct {
	Trips       int                  `json:"trips"`
	ByHour      map[string]int       `json:"byHour"`
	ByBorough   map[string]int       `json:"byBorough"`
	JoinedHours map[string][]float64 `json:"joinedHours"`
	Zone132     int                  `json:"zone132"`
	Replicas    []string             `json:"replicas"`
}

var failures []string

func check(name string, ok bool, detail string) {
	if ok {
		fmt.Printf("ok    %s\n", name)
		return
	}
	fmt.Printf("FAIL  %s   %s\n", name, detail)
	failures = append(failures, name)
}

func main() {
	address := flag.String("address", "", "the head node's address")
	expect := flag.String("expect", "", "the JSON `check.py --expect` wrote")
	flag.Parse()
	if *address == "" || *expect == "" {
		fmt.Fprintln(os.Stderr, "usage: check --address HOST:PORT --expect FILE")
		os.Exit(2)
	}

	raw, err := os.ReadFile(*expect)
	if err != nil {
		fmt.Fprintf(os.Stderr, "reading %s: %v\n", *expect, err)
		os.Exit(1)
	}
	var want expected
	if err := json.Unmarshal(raw, &want); err != nil {
		fmt.Fprintf(os.Stderr, "decoding %s: %v\n", *expect, err)
		os.Exit(1)
	}

	client, err := slate.Dial(*address, slate.Identity{
		Principal: "u64:1", Tenant: "u64:1", Roles: []string{"app"},
	})
	if err != nil {
		fmt.Fprintf(os.Stderr, "dialing %s: %v\n", *address, err)
		os.Exit(1)
	}
	defer client.Close()
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
	defer cancel()
	session := client.Session()

	// --- every row is there ------------------------------------------------
	total, served, err := countAll(ctx, session)
	if err != nil {
		fmt.Fprintf(os.Stderr, "counting: %v\n", err)
		os.Exit(1)
	}
	check("every trip is visible to the Go client too",
		total == want.Trips, fmt.Sprintf("%d against %d", total, want.Trips))
	check("and the response names the view that served it",
		served != nil && contains(want.Replicas, served.Replica),
		fmt.Sprintf("%+v", served))

	// --- a computed column, grouped ---------------------------------------
	hours := slate.Query{
		Table:   "trips",
		Compute: []slate.Scalar{slate.Extract(slate.Hour, slate.Col(tripPickupTime))},
	}
	byHour, err := groupCounts(ctx, session, hours, slate.Grouping{
		GroupBy:    []slate.Column{slate.Computed0(0)},
		Aggregates: []slate.Aggregate{slate.Count()},
	})
	if err != nil {
		fmt.Fprintf(os.Stderr, "grouping by hour: %v\n", err)
		os.Exit(1)
	}
	check("trips per hour of day agree with the Python fold",
		sameCounts(byHour, want.ByHour), diffCounts(byHour, want.ByHour))

	// --- a join, with the key and an aggregate both on the left -----------
	b := slate.NewJoin()
	trips := b.Add(slate.JoinInput{Table: "trips"})
	b.Add(slate.JoinInput{
		Table: "zones",
		On:    []slate.On{{Earlier: slate.At(trips, tripPickupZone), Own: zoneID}},
	})
	join := b.Query()
	join = slate.JoinQuery{
		Inputs: join.Inputs,
		Compute: []slate.Scalar{
			slate.Extract(slate.Hour, slate.Ref(slate.At(trips, tripPickupTime))),
		},
	}
	stream, err := session.AggregateJoin(ctx, join, slate.Grouping{
		GroupBy: []slate.Column{slate.JoinComputed(0)},
		Aggregates: []slate.Aggregate{
			slate.Count(), slate.AvgOf(slate.At(trips, tripFare)),
		},
	})
	if err != nil {
		fmt.Fprintf(os.Stderr, "grouping the join: %v\n", err)
		os.Exit(1)
	}
	groups, err := stream.Collect()
	if err != nil {
		fmt.Fprintf(os.Stderr, "draining the join: %v\n", err)
		os.Exit(1)
	}
	joined := map[string][]float64{}
	for _, group := range groups {
		hour, ok := group.Key[0].(slate.Int)
		if !ok {
			fmt.Fprintf(os.Stderr, "an hour came back as %v\n", group.Key[0])
			os.Exit(1)
		}
		count, okCount := group.Values[0].(slate.Uint)
		average, okAvg := group.Values[1].(slate.Float)
		if !okCount || !okAvg {
			fmt.Fprintf(os.Stderr, "aggregates came back as %v\n", group.Values)
			os.Exit(1)
		}
		joined[fmt.Sprint(int64(hour))] = []float64{float64(count), float64(average)}
	}
	check("the hour and the average fare, both from `trips`, across a join",
		sameJoined(joined, want.JoinedHours), diffJoined(joined, want.JoinedHours))

	// --- a group key on the right side of the join ------------------------
	rb := slate.NewJoin()
	rtrips := rb.Add(slate.JoinInput{Table: "trips"})
	rzones := rb.Add(slate.JoinInput{
		Table: "zones",
		On:    []slate.On{{Earlier: slate.At(rtrips, tripPickupZone), Own: zoneID}},
	})
	boroughStream, err := session.AggregateJoin(ctx, rb.Query(), slate.Grouping{
		GroupBy:    []slate.Column{slate.At(rzones, zoneBorough)},
		Aggregates: []slate.Aggregate{slate.Count()},
	})
	if err != nil {
		fmt.Fprintf(os.Stderr, "grouping by borough: %v\n", err)
		os.Exit(1)
	}
	boroughGroups, err := boroughStream.Collect()
	if err != nil {
		fmt.Fprintf(os.Stderr, "draining the borough grouping: %v\n", err)
		os.Exit(1)
	}
	byBorough := map[string]int{}
	for _, group := range boroughGroups {
		name, ok := group.Key[0].(slate.String)
		count, okCount := group.Values[0].(slate.Uint)
		if !ok || !okCount {
			fmt.Fprintf(os.Stderr, "a borough group is %v / %v\n", group.Key[0], group.Values[0])
			os.Exit(1)
		}
		byBorough[string(name)] = int(count)
	}
	check("grouping by the right table's borough agrees with the fold",
		sameCounts(byBorough, want.ByBorough), diffCounts(byBorough, want.ByBorough))

	// --- the index answers without reading a row --------------------------
	covering := slate.Query{
		Table:   "trips",
		Filter:  ptr(slate.Eq(tripPickupZone, slate.Uint(132))),
		Columns: []slate.Ordinal{tripPickupZone},
	}
	plan, err := session.Explain(ctx, covering)
	if err != nil {
		fmt.Fprintf(os.Stderr, "explaining: %v\n", err)
		os.Exit(1)
	}
	check("an index answers the covering query without touching a row",
		plan.IndexOnly && contains([]string{plan.Display}, plan.Display) &&
			hasSubstring(plan.Display, "by_pickup_zone"),
		fmt.Sprintf("%q index_only=%v", plan.Display, plan.IndexOnly))

	counted := slate.Query{
		Table:  "trips",
		Filter: ptr(slate.Eq(tripPickupZone, slate.Uint(132))),
	}
	inZone, _, err := aggregateOne(ctx, session, counted,
		slate.Grouping{Aggregates: []slate.Aggregate{slate.Count()}})
	if err != nil {
		fmt.Fprintf(os.Stderr, "counting the zone: %v\n", err)
		os.Exit(1)
	}
	check("the indexed count is the fold's count",
		inZone == want.Zone132, fmt.Sprintf("%d against %d", inZone, want.Zone132))

	fmt.Println()
	if len(failures) > 0 {
		fmt.Printf("%d failed: %v\n", len(failures), failures)
		os.Exit(1)
	}
	fmt.Println("the Go client agrees with the Python fold, through the same socket")
}

func ptr(e slate.Expr) *slate.Expr { return &e }

func contains(haystack []string, needle string) bool {
	for _, s := range haystack {
		if s == needle {
			return true
		}
	}
	return false
}

func hasSubstring(s, sub string) bool {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}

func countAll(ctx context.Context, session *slate.Session) (int, *slate.ServedBy, error) {
	return aggregateOne(ctx, session, slate.Query{Table: "trips"},
		slate.Grouping{Aggregates: []slate.Aggregate{slate.Count()}})
}

func aggregateOne(
	ctx context.Context, session *slate.Session,
	query slate.Query, grouping slate.Grouping,
) (int, *slate.ServedBy, error) {
	stream, err := session.Aggregate(ctx, query, grouping)
	if err != nil {
		return 0, nil, err
	}
	groups, err := stream.Collect()
	if err != nil {
		return 0, nil, err
	}
	if len(groups) == 0 {
		return 0, stream.ServedBy(), nil
	}
	count, ok := groups[0].Values[0].(slate.Uint)
	if !ok {
		return 0, nil, fmt.Errorf("a count came back as %v", groups[0].Values[0])
	}
	return int(count), stream.ServedBy(), nil
}

func groupCounts(
	ctx context.Context, session *slate.Session,
	query slate.Query, grouping slate.Grouping,
) (map[string]int, error) {
	stream, err := session.Aggregate(ctx, query, grouping)
	if err != nil {
		return nil, err
	}
	groups, err := stream.Collect()
	if err != nil {
		return nil, err
	}
	out := map[string]int{}
	for _, group := range groups {
		key, ok := group.Key[0].(slate.Int)
		count, okCount := group.Values[0].(slate.Uint)
		if !ok || !okCount {
			return nil, fmt.Errorf("a group is %v / %v", group.Key[0], group.Values[0])
		}
		out[fmt.Sprint(int64(key))] = int(count)
	}
	return out, nil
}

func sameCounts(got, want map[string]int) bool {
	if len(got) != len(want) {
		return false
	}
	for key, value := range want {
		if got[key] != value {
			return false
		}
	}
	return true
}

func diffCounts(got, want map[string]int) string {
	keys := make([]string, 0, len(want))
	for key := range want {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, key := range keys {
		if got[key] != want[key] {
			return fmt.Sprintf("%s: %d against %d", key, got[key], want[key])
		}
	}
	return fmt.Sprintf("%d groups against %d", len(got), len(want))
}

func sameJoined(got, want map[string][]float64) bool {
	if len(got) != len(want) {
		return false
	}
	for key, pair := range want {
		mine, ok := got[key]
		if !ok || len(mine) != 2 || mine[0] != pair[0] {
			return false
		}
		// The average is a float computed two different ways — a streaming
		// mean in the kernel, a sum over a slice in Python — so it is compared
		// with a tolerance rather than for equality.
		if math.Abs(mine[1]-pair[1]) > 1e-6 {
			return false
		}
	}
	return true
}

func diffJoined(got, want map[string][]float64) string {
	keys := make([]string, 0, len(want))
	for key := range want {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, key := range keys {
		mine, ok := got[key]
		if !ok {
			return fmt.Sprintf("%s: missing", key)
		}
		if len(mine) != 2 || mine[0] != want[key][0] ||
			math.Abs(mine[1]-want[key][1]) > 1e-6 {
			return fmt.Sprintf("%s: %v against %v", key, mine, want[key])
		}
	}
	return fmt.Sprintf("%d groups against %d", len(got), len(want))
}

File: examples/deployed/node/src/check.ts
/**
 * Ask the deployed head node the same questions the Python check asks, through
 * the TypeScript SDK, and require the same answers.
 *
 * The oracle stays in Python. `check.py --expect` writes what its fold
 * computed and this reads it, rather than decoding the packed trip file again:
 * three decoders of one file would be three places to be wrong, and a
 * common-mode error in them would agree with itself. What is checked here is
 * the *client* — that a TypeScript caller building the same request over the
 * same socket gets the same answer as a Python one.
 *
 * Which matters because the two build requests differently. `BigInt` versus
 * `int`, a `Column` with a `kind` versus a `ColumnRef` with a oneof, `sort`
 * keys that carry both a `column` and a `ref`. Every one of those is a place a
 * client could send something subtly different, and until this file the
 * deployed example exercised exactly one SDK.
 */

import { readFileSync } from "node:fs";
import { parseArgs } from "node:util";

import {
  at,
  avgOf,
  Client,
  col,
  computed0,
  count,
  eq,
  extract,
  joinComputed,
  newJoin,
  ref,
  uint,
  type Grouping,
  type Identity,
  type Session,
} from "@slate-orm/client";

/** What `check.py --expect` wrote. */
interface Expected {
  readonly trips: number;
  readonly byHour: Record<string, number>;
  readonly byBorough: Record<string, number>;
  readonly joinedHours: Record<string, [number, number]>;
  readonly zone132: number;
  readonly replicas: string[];
  /** The sequence the Python arm pinned its reads to. */
  readonly pinnedSequence: number;
}

// The ordinals `head.toml` declares. Written out rather than fetched, which is
// this client's arrangement too: it names a table and an ordinal.
const TRIP_PICKUP_ZONE = 1;
const TRIP_PICKUP_TIME = 3;
const TRIP_FARE = 7;
const ZONE_ID = 0;
const ZONE_BOROUGH = 1;

const failures: string[] = [];

function check(name: string, ok: boolean, detail = ""): void {
  if (ok) {
    console.log(`ok    ${name}`);
    return;
  }
  console.log(`FAIL  ${name}   ${detail}`);
  failures.push(name);
}

/** One aggregate over one table, as a number, with the view that served it. */
async function aggregateOne(
  session: Session,
  table: string,
  grouping: Grouping,
): Promise<{ value: number; servedBy: string | undefined }> {
  const stream = session.aggregate({ table }, grouping);
  const groups = await stream.collect();
  const first = groups[0];
  if (!first) return { value: 0, servedBy: stream.servedBy?.replica };
  const n = first.values[0];
  if (n?.kind !== "uint") throw new Error(`a count came back as ${n?.kind}`);
  return { value: Number(n.value), servedBy: stream.servedBy?.replica };
}

async function main(): Promise<number> {
  const { values } = parseArgs({
    options: {
      address: { type: "string" },
      expect: { type: "string" },
    },
  });
  if (!values.address || !values.expect) {
    console.error("usage: check --address HOST:PORT --expect FILE");
    return 2;
  }
  const want = JSON.parse(readFileSync(values.expect, "utf8")) as Expected;

  const identity: Identity = {
    principal: "u64:1",
    tenant: "u64:1",
    roles: ["app"],
  };
  const client = Client.connect(values.address, identity);
  const session = client.session();

  // --- pin to the snapshot the Python arm pinned to ----------------------
  //
  // This process starts with an empty watermark, so without this every read
  // below asks for nothing in particular and may be served by a replica that
  // has not finished polling the load — and then the check reports a
  // disagreement with the fold that is really a disagreement in time. It is
  // the same defect the Python arm had, and it hides better here: this runs
  // last, so the replicas have had longest and the luck holds most often.
  //
  // `observe` rather than a per-read argument because this client does not
  // have one — Python's `Client.query(freshness=…)` has no counterpart in
  // TypeScript or Go, where the session watermark is the whole mechanism.
  // `observe` is documented for carrying a position between sessions, which
  // is precisely what `pinnedSequence` is.
  check(
    "the Python arm named a sequence to pin to",
    want.pinnedSequence > 0,
    `pinnedSequence ${want.pinnedSequence}`,
  );
  session.observe(BigInt(want.pinnedSequence));

  // --- every row is there ------------------------------------------------
  const total = await aggregateOne(session, "trips", { aggregates: [count()] });
  check(
    "every trip is visible to the TypeScript client too",
    total.value === want.trips,
    `${total.value} against ${want.trips}`,
  );
  check(
    "and the response names the view that served it",
    total.servedBy !== undefined && want.replicas.includes(total.servedBy),
    `${total.servedBy}`,
  );

  // --- a computed column, grouped ---------------------------------------
  const byHour: Record<string, number> = {};
  for (const group of await session
    .aggregate(
      {
        table: "trips",
        compute: [extract("hour", col(TRIP_PICKUP_TIME))],
      },
      { groupBy: [computed0(0)], aggregates: [count()] },
    )
    .collect()) {
    const key = group.key[0];
    const n = group.values[0];
    if (key?.kind !== "int" || n?.kind !== "uint") {
      throw new Error(`a group is ${key?.kind} / ${n?.kind}`);
    }
    byHour[String(key.value)] = Number(n.value);
  }
  check(
    "trips per hour of day agree with the Python fold",
    sameCounts(byHour, want.byHour),
    firstDifference(byHour, want.byHour),
  );

  // --- a join, with the key and an aggregate both on the left -----------
  const b = newJoin();
  const trips = b.add({ table: "trips" });
  b.add({
    table: "zones",
    on: [{ earlier: at(trips, TRIP_PICKUP_ZONE), own: ZONE_ID }],
  });
  const joined: Record<string, [number, number]> = {};
  for (const group of await session
    .aggregateJoin(
      {
        ...b.query(),
        compute: [extract("hour", ref(at(trips, TRIP_PICKUP_TIME)))],
      },
      {
        groupBy: [joinComputed(0)],
        aggregates: [count(), avgOf(at(trips, TRIP_FARE))],
      },
    )
    .collect()) {
    const key = group.key[0];
    const n = group.values[0];
    const mean = group.values[1];
    if (key?.kind !== "int" || n?.kind !== "uint" || mean?.kind !== "float") {
      throw new Error(`a joined group is ${key?.kind} / ${n?.kind} / ${mean?.kind}`);
    }
    joined[String(key.value)] = [Number(n.value), mean.value];
  }
  check(
    "the hour and the average fare, both from `trips`, across a join",
    sameJoined(joined, want.joinedHours),
    JSON.stringify(Object.entries(joined).slice(0, 2)),
  );

  // --- a group key on the right side of the join ------------------------
  const rb = newJoin();
  const rtrips = rb.add({ table: "trips" });
  const rzones = rb.add({
    table: "zones",
    on: [{ earlier: at(rtrips, TRIP_PICKUP_ZONE), own: ZONE_ID }],
  });
  const byBorough: Record<string, number> = {};
  for (const group of await session
    .aggregateJoin(rb.query(), {
      groupBy: [at(rzones, ZONE_BOROUGH)],
      aggregates: [count()],
    })
    .collect()) {
    const key = group.key[0];
    const n = group.values[0];
    if (key?.kind !== "string" || n?.kind !== "uint") {
      throw new Error(`a borough group is ${key?.kind} / ${n?.kind}`);
    }
    byBorough[key.value] = Number(n.value);
  }
  check(
    "grouping by the right table's borough agrees with the fold",
    sameCounts(byBorough, want.byBorough),
    firstDifference(byBorough, want.byBorough),
  );

  // --- the index answers without reading a row --------------------------
  const plan = await session.explain({
    table: "trips",
    filter: eq(TRIP_PICKUP_ZONE, uint(132n)),
    columns: [TRIP_PICKUP_ZONE],
  });
  check(
    "an index answers the covering query without touching a row",
    plan.indexOnly && plan.display.includes("by_pickup_zone"),
    `${plan.display} indexOnly=${plan.indexOnly}`,
  );

  const inZone = await session
    .aggregate(
      { table: "trips", filter: eq(TRIP_PICKUP_ZONE, uint(132n)) },
      { aggregates: [count()] },
    )
    .collect();
  const first = inZone[0]?.values[0];
  const counted = first?.kind === "uint" ? Number(first.value) : -1;
  check(
    "the indexed count is the fold's count",
    counted === want.zone132,
    `${counted} against ${want.zone132}`,
  );

  client.close();
  console.log();
  if (failures.length) {
    console.log(`${failures.length} failed: ${failures.join(", ")}`);
    return 1;
  }
  console.log("the TypeScript client agrees with the Python fold, through the same socket");
  return 0;
}

function sameCounts(got: Record<string, number>, want: Record<string, number>): boolean {
  const keys = Object.keys(want);
  return (
    Object.keys(got).length === keys.length && keys.every((key) => got[key] === want[key])
  );
}

function firstDifference(got: Record<string, number>, want: Record<string, number>): string {
  for (const key of Object.keys(want).sort()) {
    if (got[key] !== want[key]) return `${key}: ${got[key]} against ${want[key]}`;
  }
  return `${Object.keys(got).length} groups against ${Object.keys(want).length}`;
}

function sameJoined(
  got: Record<string, [number, number]>,
  want: Record<string, [number, number]>,
): boolean {
  const keys = Object.keys(want);
  if (Object.keys(got).length !== keys.length) return false;
  return keys.every((key) => {
    const mine = got[key];
    const theirs = want[key]!;
    // The average is a float computed two different ways — a streaming mean in
    // the kernel, a sum over a list in Python — so it is compared with a
    // tolerance rather than for equality.
    return mine !== undefined && mine[0] === theirs[0] && Math.abs(mine[1] - theirs[1]) < 1e-6;
  });
}

process.exitCode = await main();
